### PR TITLE
add logs and fix error reasons

### PR DESCRIPTION
**Pull request by OrenZak**
#### Main purpose:
* Add more logs and events
* Fix error reasons null and blanks

### CODE REVIEW

File: core/src/main/java/com/kin/ecosystem/core/bi/events/MaxWalletsExceededError.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+import com.google.gson.annotations.Expose;
+import com.google.gson.annotations.SerializedName;
+
+
+/**
+ * when a user exceeded max wallet limit
+ * 
+ */
+public class MaxWalletsExceededError implements Event {
+    public static final String EVENT_NAME = "max_wallets_exceeded_error";
```
Comment (doodyparizada) [inline core/src/main/java/com/kin/ecosystem/core/bi/events/MaxWalletsExceededError.java:19]: @neritBA Did you define this new metric?

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSource.java [hunk 2]
```diff
@@ -213,8 +211,6 @@ void sendTransaction(@NonNull String publicAddress, @NonNull BigDecimal amount,
 
 		void getBlockchainVersion(@NonNull final Callback<KinSdkVersion, ApiException> callback);
 
-		MigrationInfo getMigrationInfo(String publicAddress)  throws ApiException; // synced and blocking
```
Comment (doodyparizada) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSource.java]: Why are you removing this?

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java [hunk 3]
```diff
@@ -461,6 +465,11 @@ public void deleteAccount(int accountIndex) throws DeleteAccountException {
 		}
 	}
 
+	@Override
+	public ArrayList<String> getWalletsAddress(String kinUserId) {
+		return local.getWalletsAddress(kinUserId);
```
Comment (doodyparizada) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java]: Can you get the entire list of wallet addresses for all users on this device?

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceLocal.java [hunk 4]
```diff
@@ -58,6 +58,11 @@ public String getLastWalletAddress(final String kinUserId) {
 		return wallets.size() > 0 ? wallets.get(wallets.size() - 1) : null;
 	}
 
+	@Override
+	public ArrayList<String> getWalletsAddress(final String kinUserId) {
```
Comment (doodyparizada) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceLocal.java]: `getWalletAddresses`

File: core/src/main/java/com/kin/ecosystem/core/util/DateUtil.java [hunk 5]
```diff
@@ -13,8 +13,10 @@
     private static DateFormat utcDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm", Locale.US);
     private static TimeZone utcTimeZone = TimeZone.getTimeZone("UTC");
 
+    private static DateFormat utcDateFormatExtended = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US);
 	static {
 		utcDateFormat.setTimeZone(utcTimeZone);
+        utcDateFormatExtended.setTimeZone(utcTimeZone);
```
Comment (doodyparizada) [inline core/src/main/java/com/kin/ecosystem/core/util/DateUtil.java:22]: code alignment issue?

File: sdk/src/main/java/com/kin/ecosystem/Kin.java [hunk 6]
```diff
@@ -324,7 +333,7 @@ public void run() {
 	}
 
 	private static void sendLoginFailed(final KinEcosystemException exception, final KinCallback<Void> loginCallback) {
-		eventLogger.send(UserLoginFailed.create(getMessage(exception)));
+		eventLogger.send(UserLoginFailed.create(ErrorUtil.getMessage(exception, "User login failed with unknown exception ")));
```
Comment (doodyparizada) [inline sdk/src/main/java/com/kin/ecosystem/Kin.java]: extra space is extra `exception "`

File: sdk/src/main/java/com/kin/ecosystem/Kin.java [hunk 7]
```diff
@@ -259,6 +263,11 @@ public void onResponse(Boolean response) {
 
 						@Override
 						public void onFailure(KinEcosystemException exception) {
+							if (exception.getCode() == ServiceException.MAX_WALLETS_EXCEEDED) {
```
Comment (doodyparizada) [inline sdk/src/main/java/com/kin/ecosystem/Kin.java]: Do you have something that tests this?

File: core/src/main/java/com/kin/ecosystem/core/util/DateUtil.java (hunk 5 above)
Comment (liorama) [inline core/src/main/java/com/kin/ecosystem/core/util/DateUtil.java:22]: Is this intentionally indented this way?

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSource.java (hunk 2 above)
Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSource.java]: not used

File: core/src/main/java/com/kin/ecosystem/core/bi/events/MaxWalletsExceededError.java (hunk 1 above)
Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/bi/events/MaxWalletsExceededError.java:19]: yes she did

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java (hunk 3 above)
Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java]: yes

File: core/src/main/java/com/kin/ecosystem/core/util/DateUtil.java (hunk 5 above)
Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/util/DateUtil.java:22]: only on github

File: core/src/main/java/com/kin/ecosystem/core/util/DateUtil.java [hunk 8]
```diff
@@ -13,8 +13,10 @@
     private static DateFormat utcDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm", Locale.US);
     private static TimeZone utcTimeZone = TimeZone.getTimeZone("UTC");
 
+    private static DateFormat utcDateFormatExtended = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US);
```
Comment (liorama) [inline core/src/main/java/com/kin/ecosystem/core/util/DateUtil.java]: I think the `Z` and `Locale.US` conflict, It's either UTC or US.

File: sdk/src/main/java/com/kin/ecosystem/Kin.java (hunk 7 above)
Comment (OrenZak) [reply sdk/src/main/java/com/kin/ecosystem/Kin.java]: only manual at the moment

**Comment by liorama** [comment]
@oren-a Did you add the error message for `MigrationFailedException`?

**Comment by oren-a** [comment]
> @oren-a Did you add the error message for `MigrationFailedException`?

@liorama - I guess you meant @OrenZak

File: core/src/main/java/com/kin/ecosystem/core/util/DateUtil.java (hunk 5 above)
Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/util/DateUtil.java:22]: github issue somehow

**Comment by OrenZak** [comment]
> @oren-a Did you add the error message for `MigrationFailedException`?

yes

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java [hunk 9]
```diff
@@ -466,8 +466,21 @@ public void deleteAccount(int accountIndex) throws DeleteAccountException {
...
 
 	@Override
-	public ArrayList<String> getWalletsAddress(String kinUserId) {
-		return local.getWalletsAddress(kinUserId);
+	public ArrayList<String> getWalletAddresses(String kinUserId) {
+		return local.getWalletAddresses(kinUserId);
+	}
+
+	@Override
+	public ArrayList<String> getAllWalletAddresses() {
```
Comment (doodyparizada) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java]: general question: why do you return an ArrayList and not just a List?

**Review by doodyparizada — APPROVED** [review]
looks good

**Pull request merged**